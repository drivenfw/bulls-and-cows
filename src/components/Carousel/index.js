import React from 'react'
import styled from 'styled-components'
import { CSSTransition, TransitionGroup } from 'react-transition-group'


const StyledCarousel = styled.div`
  box-sizing: border-box;
  width: 35px;
  height: 60px;
  border: 1px solid grey;
  border-radius: 3px;
  box-shadow: 0 0 5px 3px lightgrey;
  font-size: 2em;
  position: relative;
  overflow: hidden;

  & .turn-down-enter {
    top: -100%;
    transition: top 500ms cubic-bezier(0, 0.7, 0.25, 1.19);
  }

  & .turn-down-enter-active {
    top: 0;
  }

  & .turn-down-exit {
    top: 0;
    transition: top 500ms cubic-bezier(0, 0.7, 0.25, 1.19);
  }

  & .turn-down-exit-active {
    top: 100%;
  }

  & .turn-up-enter {
    top: 100%;
    transition: top 500ms cubic-bezier(0, 0.7, 0.25, 1.19);
  }

  & .turn-up-enter-active {
    top: 0;
  }

  & .turn-up-exit {
    top: 0;
    transition: top 500ms cubic-bezier(0, 0.7, 0.25, 1.19);
  }

  & .turn-up-exit-active {
    top: -100%;
  }
`

export const Value = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  position: absolute;
`

const Carousel = ({ children, className, direction }) => 
  <StyledCarousel className={className}>
    <TransitionGroup>
      {React.Children.map(children, child => (
        <CSSTransition
          timeout={500}
          classNames={`turn-${direction}`}
          mountOnEnter
          unmountOnExit
        >
          {child}
        </CSSTransition>
      ))}
    </TransitionGroup>
  </StyledCarousel>

export default Carousel

