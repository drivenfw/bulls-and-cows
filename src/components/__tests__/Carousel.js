import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import Carousel, { Value } from 'components/Carousel'


configure({ adapter: new Adapter() })

describe('<Carousel /> snapshot test', () => {
  it('should match the snapshot', () => {
    const tree = shallow(
      <Carousel direction="up">
        <Value>One</Value>
      </Carousel>
    )

    expect(toJson(tree)).toMatchSnapshot()
  })
})

