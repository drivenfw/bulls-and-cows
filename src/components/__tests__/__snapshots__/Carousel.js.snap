// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> snapshot test should match the snapshot 1`] = `
<styled.div>
  <TransitionGroup
    childFactory={[Function]}
    component="div"
  >
    <CSSTransition
      classNames="scroll-up"
      key=".0"
      mountOnEnter={true}
      timeout={500}
      unmountOnExit={true}
    >
      <styled.div>
        One
      </styled.div>
    </CSSTransition>
  </TransitionGroup>
</styled.div>
`;
